tidy3d/components/source.py

```diff
@@ -373,18 +388,34 @@ def amp_time(self, time: float) -> complex:
         complex
             Complex-valued source amplitude at that time.
         """
+
         if self.source_time_dataset is None:
             return None
-        times = self.source_time_dataset.values.coords["t"].values.squeeze()
-        if isinstance(time, float):
-            if time < min(times) or time > max(times):
-                return self.source_time_dataset.values.sel(t=time, method="nearest").to_numpy()
-            return self.source_time_dataset.values.interp(t=time).to_numpy()
-        val = np.zeros(len(time), dtype=complex)
-        mask = (time < min(times)) | (time > max(times))
-        val[mask] = self.source_time_dataset.values.sel(t=time[mask], method="nearest").to_numpy()
-        val[~mask] = self.source_time_dataset.values.interp(t=time[~mask]).to_numpy()
-        return val
+
+        # make time a numpy array for uniform handling
+        times = np.array([time] if isinstance(time, float) else time)
+        data_times = self.source_time_dataset.values.coords["t"].values.squeeze()
+
+        # shift time
+        twidth = 1.0 / (2 * np.pi * self.fwidth)
+        time_shifted = times - self.offset * twidth
+
+        # mask times that are out of range
+        mask = (time_shifted < min(data_times)) | (time_shifted > max(data_times))
+
+        # get envelope
+        envelope = np.zeros(len(time_shifted), dtype=complex)
+        values = self.source_time_dataset.values
+        envelope[mask] = values.sel(t=time_shifted[mask], method="nearest").to_numpy()
+        envelope[~mask] = values.interp(t=time_shifted[~mask]).to_numpy()
+
+        # modulation, phase, amplitude
```

Comment:
I think it would be useful to have a formula of the pulse amplitude as a function of these variables in the docstring description